cadc-uws-server/src/main/java/ca/nrc/cadc/uws/web/JobAction.java

```diff
@@ -0,0 +1,186 @@
+    protected JobManager jobManager;
+    
+    // JobAttribute has strings used in XML document; the REST child resource names and update param names can differ
+    protected static final Map<String,JobAttribute> CHILD_RESOURCE_NAMES = new HashMap<String,JobAttribute>();
+    protected static final Map<JobAttribute,String> CHILD_PARAM_NAMES = new HashMap<JobAttribute,String>();
+    
+    static {
+        CHILD_RESOURCE_NAMES.put("phase", JobAttribute.EXECUTION_PHASE);
+        CHILD_RESOURCE_NAMES.put("executionduration", JobAttribute.EXECUTION_DURATION);
+        CHILD_RESOURCE_NAMES.put("destruction", JobAttribute.DESTRUCTION_TIME);
+        CHILD_RESOURCE_NAMES.put("quote", JobAttribute.QUOTE);
+        CHILD_RESOURCE_NAMES.put("owner", JobAttribute.OWNER_ID); 
+        CHILD_RESOURCE_NAMES.put("parameters", JobAttribute.PARAMETERS);
+        
+        CHILD_PARAM_NAMES.put(JobAttribute.EXECUTION_PHASE, JobAttribute.EXECUTION_PHASE.getValue().toUpperCase());
+        CHILD_PARAM_NAMES.put(JobAttribute.EXECUTION_DURATION, JobAttribute.EXECUTION_DURATION.getValue().toUpperCase());
+        CHILD_PARAM_NAMES.put(JobAttribute.DESTRUCTION_TIME, JobAttribute.DESTRUCTION_TIME.getValue().toUpperCase());
+        CHILD_PARAM_NAMES.put(JobAttribute.QUOTE, JobAttribute.QUOTE.getValue().toUpperCase());
+    }
+    
+    public JobAction() { 
+        super();
+    }
+    
+    /**
+     * Initialisation that cannot be performed in the constructor. This method must
+     * be called at the start of doAction in all subclasses.
+     */
+    protected void init() {
+        if (jobManager == null) {
+            String jndiKey = componentID + ".jobManager"; // see AsyncServlet
+            try {
+                Context ctx = new InitialContext();
+                this.jobManager = (JobManager) ctx.lookup(jndiKey);
+                if (jobManager != null) {
+                    log.debug("found: " + jndiKey +"=" + jobManager.getClass().getName());
+                } else {
+                    log.error("BUG: failed to find " + jndiKey + " via JNDI");
+                }
+            } catch (NamingException ex) {
+                log.error("BUG: failed to find " + jndiKey + " via JNDI", ex);
+            }
+        }
+    }
+    
+    private void initTarget() {
+        // [jobID[/jobField[/resultID]]]
+        if (jobID == null) {
+            String path = syncInput.getPath();
+            if (path != null) {
+                String[] parts = path.split("/");
+                if (parts.length > 0) {
+                    this.jobID = parts[0];
+                    if (parts.length > 1) {
+                        this.jobField = parts[1];
+                        if (parts.length == 3) {
+                            this.resultID = parts[2];
+                        }
+                    }
+                }
+            }
+        }
+        // TODO: validation
+        log.debug("init: " + jobID + " " + jobField + " " + resultID);
+    }
```

Comment:
it is a lazy init because action classes only have no-arg constructor so yes: have to wait for some set methods to be called by RestAction. the initTarget method itself only does work once (skipped once jobID is set) except in the case of a get or post to the job list, but it didn't seem worth tracking that it had already run when in those cases it is just a single extra comparison (line 141) before it returns.

I suppose a solution would be to have a protected no-op init method that RestServlet would call after all setup and before calling run()... I think I would prefer to port something else like cadc-vos-server from restlet to cadc-rest and refactor at that point.